os/arch/arm/src/armv7-a/arm_schedulesigaction.c

```diff
@@ -0,0 +1,444 @@
+ *
+ * Assumptions:
+ *   Called from critical section
+ *
+ ****************************************************************************/
+
+#ifndef CONFIG_SMP
+void up_schedule_sigaction(struct tcb_s *tcb, sig_deliver_t sigdeliver)
+{
+  sinfo("tcb=%p sigdeliver=%p\n", tcb, sigdeliver);
+
+  /* Refuse to handle nested signal actions */
+
+  if (!tcb->xcp.sigdeliver)
+    {
+      /* First, handle some special cases when the signal is being delivered
+       * to task that is currently executing on this CPU.
+       */
+
+      sinfo("rtcb=%p CURRENT_REGS=%p\n", this_task(), CURRENT_REGS);
+
+      if (tcb == this_task())
+        {
+          /* CASE 1:  We are not in an interrupt handler and a task is
+           * signaling itself for some reason.
+           */
+
+          if (!CURRENT_REGS)
+            {
+              /* In this case just deliver the signal now.
+               * REVISIT:  Signal handler will run in a critical section!
+               */
+
+              sigdeliver(tcb);
+            }
+
+          /* CASE 2:  We are in an interrupt handler AND the interrupted
+           * task is the same as the one that must receive the signal, then
+           * we will have to modify the return state as well as the state
+           * in the TCB.
+           *
+           * Hmmm... there looks like a latent bug here: The following logic
+           * would fail in the strange case where we are in an interrupt
+           * handler, the thread is signaling itself, but a context switch
+           * to another task has occurred so that CURRENT_REGS does not
+           * refer to the thread of this_task()!
+           */
+
+          else
+            {
+              /* Save the return lr and cpsr and one scratch register
+               * These will be restored by the signal trampoline after
+               * the signals have been delivered.
+               */
+
+              tcb->xcp.sigdeliver     = sigdeliver;
+
+              /* And make sure that the saved context in the TCB is the same
+               * as the interrupt return context.
+               */
+
+              arm_savestate(tcb->xcp.saved_regs);
+
+              /* Duplicate the register context.  These will be
+               * restored by the signal trampoline after the signal has been
+               * delivered.
+               */
+
+              CURRENT_REGS            = (void *)
+                                        ((uint32_t)CURRENT_REGS -
+                                         (uint32_t)XCPTCONTEXT_SIZE);
+              memcpy((uint32_t *)CURRENT_REGS, tcb->xcp.saved_regs,
+                     XCPTCONTEXT_SIZE);
+
+              CURRENT_REGS[REG_SP]    = (uint32_t)CURRENT_REGS +
+                                        (uint32_t)XCPTCONTEXT_SIZE;
+
+              /* Then set up to vector to the trampoline with interrupts
+               * disabled
+               */
+
+              CURRENT_REGS[REG_PC]    = (uint32_t)up_sigdeliver;
```

Comment:
Could you check please?
Is up_sigdeliver() used in Armv7-a? or arm_sigdeliver()